⬆️(@typescript-eslint.v7.7.0) replace the deprecated rules

kimk113 opened the pull request:
### Contents
[typescript-eslint v7.7.0](https://typescript-eslint.io/rules/) 도큐먼트 참조, deprecated 된 룰을 삭제합니다. 

```
- @typescript-eslint/no-duplicate-imports
- @typescript-eslint/no-implicit-any-catch
- @typescript-eslint/no-parameter-properties
```

laziel commented:
> Previously deprecated rules are deleted:
> - `@typescript-eslint/no-duplicate-imports`, replaced by `imports/no-duplicates`
> - `@typescript-eslint/no-implicit-any-catch`, replaced by the TSConfig option [useUnknownInCatchVariables](https://www.typescriptlang.org/tsconfig#useUnknownInCatchVariables)
> - `@typescript-eslint/no-parameter-properties`, replaced by `@typescript-eslint/parameter-properties`

https://typescript-eslint.io/blog/announcing-typescript-eslint-v6/

laziel commented:
`@typescript-eslint/no-duplicate-imports`:
이미 `no-duplicate-imports` 규칙이 있어서 `imports/no-duplicates`로 대치하지 않고 그냥 삭제하신 것 같은데 살짝 동작 방식이 다르다고 합니다. 검토 부탁드립니다.
https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-duplicates.md

`@typescript-eslint/no-parameter-properties` 규칙은 `@typescript-eslint/parameter-properties`로 교체되었다고 하는데요 그냥 삭제해도 괜찮을까요?
잘 쓰지는 않고 있는 규칙이라서 크게 영향은 없을 것 같은데 단순 버전업이 아닌 규칙 변경에 해당하는 부분이라 기록해둬야 할 것 같습니다.
https://typescript-eslint.io/rules/parameter-properties/

kimk113 commented:
replaced 안내가 있는 규칙은 삭제 대신 교체로 수정했습니다.

`@typescript-eslint` 버전 5. 를 사용중인 프로젝트는 `@typescript-eslint/parameter-properties` 신규룰이 없어 해당 패키지 v1.1.2 와 호환이 안될 수 있습니다.  

publish 해서 적용 전까지는 검증이 어렵네요..
그 외 parserOptions 변경사항은 건드리지 않았습니다. 개별 프로젝트 버전에 따라 알아서 덮어쓰도록..

liebebe reviewed (approved):
> publish 해서 적용 전까지는 검증이 어렵네요..
그 외 parserOptions 변경사항은 건드리지 않았습니다. 개별 프로젝트 버전에 따라 알아서 덮어쓰도록..

npm link 이용하면 패키지를 리모트 환경에 publish 하기 전에 로컬에서 테스트 해볼 수 있습니다. https://docs.npmjs.com/cli/v10/commands/npm-link

kimk113 commented:
npm link 로 테스트시 다음과 같은 현상이 있는데 혹시 아시나요?
파서 지정은 이전에도 있던 내용인데 스콥이 달라져서 그런지 엉뚱한데서 찾는 모양인데요..

<img width="1113" alt="스크린샷 2024-04-17 오후 1 23 54" src="https://github.com/naverwhale/eslint-config-whale-web/assets/30401541/609da6c5-3164-4ee9-88dc-0348c7f3786c">

liebebe commented:
> npm link 로 테스트시 다음과 같은 현상이 있는데 혹시 아시나요?
> 파서 지정은 이전에도 있던 내용인데 스콥이 달라져서 그런지 엉뚱한데서 찾는 모양인데요..

파서를 이름으로만 지정하면 base config 디렉토리에서 찾는가보네요. require.resolve 이용하는 방식으로 해보시겠어요?
https://github.com/eslint/eslint/issues/16942

```diff
-parser: '@typescript-eslint/parser',
+parser: require.resolve('@typescript-eslint/parser'),
```

Commit:
⬆️(@typescript-eslint.v7.7.0) replace deprecated rule

- https://typescript-eslint.io/blog/announcing-typescript-eslint-v6/#rule-breaking-changes

Commit:
🔖️ v1.1.2

kimk113 commented:
- import 플러그인을 추가하고, 교체되는 룰의 오타를 수정했습니다. `imports/no-duplicates` -> `import/no-duplicates`

---


> 파서를 이름으로만 지정하면 base config 디렉토리에서 찾는가보네요. require.resolve 이용하는 방식으로 해보시겠어요? [eslint/eslint#16942](https://github.com/eslint/eslint/issues/16942)
> 
> ```diff
> -parser: '@typescript-eslint/parser',
> +parser: require.resolve('@typescript-eslint/parser'),
> ```

이것도 동작하지 않습니다. 
여러 글을 찾긴 했는데 해결이 안되어서 정확한 원인은 모르겠네요.   

- https://typescript-eslint.io/blog/parser-options-project-true#project-services 
> Complex project setups can be difficult to get right.
For example, [typescript-eslint does not yet support Project References](https://github.com/typescript-eslint/typescript-eslint/issues/2094).

- https://github.com/microsoft/vscode-eslint/issues/1543
> The problem is caused by the TypeScript compiler now real pathing incoming paths. I think they added that in 5.x.

로컬에서 테스트는 우선 다른 방식으로 완료해서 여기서 마무리하고 머지하도록 하곘습니다. 
이건 나중에 모노레포라도 구성하게되면 다시 봐야겠네요.
리뷰 감사합니다!
